spores-pickling/src/main/scala/scala/spores/SporePicklers.scala

```diff
@@ -0,0 +1,372 @@
+    * The `Pickler`s will be used by default and the `Unpickler`s will be invoked
+    * by the `UnpicklerFetcher` defined below. Fortunately, we don't have to keep
+    * type information of `Captured` in the pickled message since we already know
+    * which is the correct tag when `unpickle` is invoked (this is kind of a hack).
+    */
+  def genSporePicklerUnpicklerTemplate[U: c.WeakTypeTag]
+      (c: blackbox.Context)(sporeType: c.Tree): c.Tree = {
+
+    import c.universe._
+
+    val utpe = weakTypeOf[U].dealias
+    debug(s"Captured type is: $utpe")
+
+    val reader = c.freshName(TermName("reader"))
+    val builder = c.freshName(TermName("builder"))
+    val picklee = c.freshName(TermName("picklee"))
+    val className = c.freshName(TermName("className"))
+    val unpickledSpore = c.freshName(TermName("spore"))
+    val unpickledCapture = c.freshName(TermName("capture"))
+    val capturedPickler = c.freshName(TermName("capturedPickler"))
+    val capturedUnpickler = c.freshName(TermName("capturedUnpickler"))
+    val picklerUnpicklerName = c.freshName(TermName("SporePicklerUnpickler"))
+
+    val utils = new PicklerUtils[c.type](c)
+    import utils.{absPicklerUnpicklerType, picklerType, unpicklerType, pbuilderType, preaderType, scalaPath}
+    import utils.{fastTypeTagType, stringType, anyType, unitType, autoRegisterType, locallyPath}
+    val resultType = tq"$absPicklerUnpicklerType[$sporeType]"
+
+    q"""
+      $locallyPath[$resultType] {
+        val $capturedPickler = implicitly[$picklerType[$utpe]]
+        val $capturedUnpickler = implicitly[$unpicklerType[$utpe]]
+
+        implicit object $picklerUnpicklerName
+            extends $absPicklerUnpicklerType[$sporeType]
+              with $autoRegisterType[$sporeType] {
+
+          def tag = implicitly[$fastTypeTagType[$sporeType]]
+
+          def pickle($picklee: $sporeType, $builder: $pbuilderType): $unitType = {
+
+            $builder.beginEntry($picklee, tag)
+            // Runtime picklers need the type, don't elide it
+            // builder.hintElidedType(tag)
+            ${utils.writeUnpicklerClassName(builder, picklerUnpicklerName)}
+            ${utils.writeClassName(builder, picklee)}
+            ${utils.writeCaptured(builder, picklee, capturedPickler, sporeType, utpe)}
+            $builder.endEntry()
+
+          }
+
+          def unpickle(tag: $stringType, $reader: $preaderType): $anyType = {
+
+            val $className = ${utils.readClassName(reader)}
+            val $unpickledSpore = ${utils.createInstance(className, sporeType)}
+
+            /* Ask for the unpickler at this point, since if we pass a var outside
+             * this scope, the unpickler will be null. This is a safe operation
+             * since such an `Unpickler` exists because we got it as an implicit */
+
+            val $capturedUnpickler = implicitly[$unpicklerType[$utpe]]
+            if($capturedUnpickler == null) println("Unpickler[" + ${utpe.toString} + "] is null")
```

Comment:
We probably don't want to keep this `println` there, do we? Perhaps throwing an exception would be better.